teps/0076-array-result-types.md

```diff
@@ -0,0 +1,563 @@
+* Must be possible to use one specific value in an array result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value at the specified index, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+Currently we treat the contents of
+[a Task's result file(s)](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results)
+as a string. The file is located at `$(results.resultName.path)` which expands to `/tekton/results/resultName`.
+In this proposal, the results section of the Task will inform the controller about whether it should interpret the
+content of the result as [json](#why-json) or as a string:
+  * If the type indicates that the result is an array, the controller will interpret the result as json and expect an
+    array. If the content is not json or not an array, the TaskRun will fail
+  * If the type indicates that the result is a string, the controller will simply use the result as is (i.e. the
+    same logic as today)
+
+Pros:
+  * Completely backwards compatible
+
+Cons:
+  * Additional complexity in that the results file could contain two different kinds of content
+
+This feature would be considered alpha and would be
+[gated by the alpha flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+
+### Why json?
+
+This proposal suggests using `json` as the format for more complex types like arrays (and in
+[TEP-0075](https://github.com/tektoncd/community/pull/479), dictionaries). Possible other formats we could use are:
+
+* `yaml` (i dont think there is a syntax like jsonpath available for yaml? i.e. it'd be harder to expose via variable
+  replacement - or we'd need to convert from yaml to json internally)
+* `protobuf` (this would dramatically increase the difficulty in emitting structured results from tasks - either we'd
+  see every Task including a step with protobuf tools in it or we'd need to include some by default)
+* `xml` (seems more verbose than we'd need and afaik in cloud native tools isn't in wide use)
+* `csv` (would also be easy to use but afaik it wouldn't support more complex types like dicts? or we'd need to add some
+  kind of syntax on top of the csv format to define more complex types)
+
+This proposal suggests using `json` because:
+* It's human-readable
+* It's possible for Tasks to write json content without needing extra tools
+* We've added json specific support to other parts of our APIs (e.g. [in triggers](https://github.com/tektoncd/triggers/blob/f2380d7d33666547ee7ed965730cf7dbc1aa5421/docs/triggerbindings.md#accessing-data-in-http-json-payloads))
+* Kubernetes supports `yaml` and `json`, so it makes sense to narrow our choices to at least those two, as opposed to
+  introducing a new format into the mix.
+
```

Comment:
@mattmoor pointed out today that another good option would be to adopt [the open api spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md) that CRDs use to day

TODO: explore as an alternative

Reply:
ah this actually makes more sense as a json schema alternative - going to add a "why json schema" section to https://github.com/tektoncd/community/pull/479 - and TL;DR it seems like Open API schemas might be a slightly better (more consistent with k8s and with what we want) approach for schemas but we don't need to cross that bridge until we expand our schema support